ast2vec/cloning.py

```diff
@@ -0,0 +1,213 @@
+    """
+    Raised when we fail to classify the source files.
+    """
+    pass
+
+
+class RepoCloner:
+    """
+    Clone repositories from provided urls / files with urls.
+    Use enry to classify files and delete redundant files if needed.
+    """
+    def __init__(self, linguist, redownload, languages=None,
+                 log_level=logging.INFO, num_threads=1):
+        self._log = logging.getLogger("repo_cloner")
+        self._log.setLevel(log_level)
+        self._languages = languages
+        self._linguist, self._is_enry = self._find_enry(linguist)
+        self._num_threads = num_threads
+        self._redownload = redownload
+
+    def clone_repo(self, git_url, ignore, target_dir):
+        """
+        Clone repository into a separate folder inside of target directory.
+
+        :param git_url: Url of Git repository.
+        :param ignore: Flag for ignoring Git failures.
+        :param target_dir: Target directory. New folder will be created inside of target_dir.
+        :return: Path to downloaded Git repository.
+        """
+        git_url = self._prepare_repo_url(git_url)
+
+        try:
+            repo_dir = self._prepare_repo_dir(git_url, target_dir, self._redownload)
+        except FileExistsError as e:
+            self._log.warning("%s already cloned, skipping.", git_url)
+            return
+
+        env = os.environ.copy()
+        env["GIT_TERMINAL_PROMPT"] = "0"
+        self._log.info("Cloning from %s...", git_url)
+
+        try:
+            subprocess.check_output(["git", "clone", "--depth=1", git_url, repo_dir],
+                                    env=env, stdin=subprocess.DEVNULL, stderr=subprocess.STDOUT)
+        except subprocess.CalledProcessError as e:
+            shutil.rmtree(repo_dir, ignore_errors=True)
+            self._log.error("Git failed to clone repo. git stderr:\n\t" +
+                            "\n\t".join(e.output.decode("utf8").split("\n")))
+            if not ignore:
+                raise e from None
+            else:
+                return
+        except Exception as e:
+            shutil.rmtree(repo_dir, ignore_errors=True)
+            self._log.error("Unknown error in RepoCloner. Git failed to clone repo.")
+            raise e from None
+
+        self._log.info("Finished cloning %s", git_url)
+        return repo_dir
+
+    def classify_repo(self, repo_dir: str) -> dict:
+        """
+        Classify files in a repository using enry.
+
+        :param repo_dir: Path to repository directory.
+        :return: enry output loaded from JSON.
+        """
+        if not self._linguist:
+            raise LinguistFailedError("Linguist is not set - cannot classify the files")
+
+        self._log.info("Classifying the files...")
+        repo_dir = os.path.abspath(repo_dir)
+        cmdline = [self._linguist]
+        if self._is_enry:
+            cmdline += ["-json", repo_dir]
+        else:
+            cmdline += [repo_dir, "--json"]
```

Comment:
I think that nothing was tested with linguist, but still, it can be used) let's keep it as it's right now